### ISSUE TITLE
Can not enable large bar on centos7.9+vega20

### ISSUE THREAD

**Issue by raykwok1150** [RadeonOpenCompute/ROCK-Kernel-Driver#109]
Greets,
I can not enbale large bar on cnetos7.9.
Here is the error log

```
[   71.290581] [drm] GPU posting now...
[   71.290994] [drm] vm size is 262144 GB, 4 levels, block size is 9-bit, fragment size is 9-bit
[   71.291001] amdgpu 0000:03:00.0: VRAM: 16368M 0x0000008000000000 - 0x00000083FEFFFFFF (16368M used)
[   71.291002] amdgpu 0000:03:00.0: GART: 512M 0x0000000000000000 - 0x000000001FFFFFFF
[   71.291003] amdgpu 0000:03:00.0: AGP: 267894784M 0x0000008400000000 - 0x0000FFFFFFFFFFFF
[   71.291008] [drm] Detected VRAM RAM=16368M, BAR=256M
[   71.291009] [drm] RAM width 4096bits HBM
[   71.291126] [drm] amdgpu: 16368M of VRAM memory ready
[   71.291128] [drm] amdgpu: 64131M of GTT memory ready.
[   71.291133] [drm] GART: num cpu pages 131072, num gpu pages 131072
[   71.291205] [drm] PCIE GART of 512M enabled (table at 0x0000008000000000).
[   71.291246] amdgpu 0000:03:00.0: irq 141 for MSI/MSI-X
```

After traced the linux kernel code, I think this the root course 
```
commit 9b6fa93aed993e866063ece43e56dafbf591fea8 (HEAD -> master)
Author: Guo Lei <raykwok1150@163.com>
Date:   Wed Feb 24 21:11:21 2021 +0800

    drm/amdkcl: Always resize BAR state.
    
    The linux kerne version of centos 7.x is 3.10.x, which is less than
    4.15.0. So, we must always resize the bar state.
    
    Tested on centos7.9 + vega20.
    
    Signed-off-by: Guo Lei <raykwok1150@163.com>

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu.h b/drivers/gpu/drm/amd/amdgpu/amdgpu.h
index 85ed160c7c81..3f12db851362 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu.h
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu.h
@@ -1219,14 +1219,7 @@ bool amdgpu_device_need_post(struct amdgpu_device *adev);
 
 void amdgpu_cs_report_moved_bytes(struct amdgpu_device *adev, u64 num_bytes,
                                  u64 num_vis_bytes);
-#ifdef AMDKCL_ENABLE_RESIZE_FB_BAR
 int amdgpu_device_resize_fb_bar(struct amdgpu_device *adev);
-#else
-static inline int amdgpu_device_resize_fb_bar(struct amdgpu_device *adev)
-{
-       return 0;
-}
-#endif
 void amdgpu_device_program_register_sequence(struct amdgpu_device *adev,
                                             const u32 *registers,
                                             const u32 array_size);
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_device.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_device.c
index 58ca17e894fa..e6c7c25f0fa5 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_device.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_device.c
@@ -925,7 +925,6 @@ void amdgpu_device_wb_free(struct amdgpu_device *adev, u32 wb)
                __clear_bit(wb, adev->wb.used);
 }
 
-#ifdef AMDKCL_ENABLE_RESIZE_FB_BAR
 /**
  * amdgpu_device_resize_fb_bar - try to resize FB BAR
  *
@@ -1000,7 +999,6 @@ int amdgpu_device_resize_fb_bar(struct amdgpu_device *adev)
 
        return 0;
 }
-#endif
 
 /*
  * GPU helpers function.
diff --git a/include/kcl/backport/kcl_drm_backport.h b/include/kcl/backport/kcl_drm_backport.h
index f32e51d37d7d..191f8ff23ac3 100644
--- a/include/kcl/backport/kcl_drm_backport.h
+++ b/include/kcl/backport/kcl_drm_backport.h
@@ -61,14 +61,6 @@ int _kcl_drm_syncobj_find_fence(struct drm_file *file_private,
 #define drm_syncobj_find_fence _kcl_drm_syncobj_find_fence
 #endif
 
-/*
- * commit d3252ace0bc652a1a244455556b6a549f969bf99
- * PCI: Restore resized BAR state on resume
- */
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
-#define AMDKCL_ENABLE_RESIZE_FB_BAR
-#endif
-
 /*
  * commit v4.10-rc3-539-g086f2e5cde74
  * drm: debugfs: Remove all files automatically on cleanup
```
Can I merge this patch to current repository?

BR
Guo Lei

**Comment by fxkamd** [comment]
I don't think this is the correct solution. There should be a proper configure.sh check implemented to check for the presence of the BAR resizing functionality in the kernel. The kernel version check that is currently used probably doesn't take into account a backport done by RedHat.

**Comment by raykwok1150** [comment]
Hi fxkamd,
Follow your tips, I have two ideas.

1, define AMDKCL_ENABLE_RESIZE_FB_BAR when OS_NAME_RHEL_7_X
The patch is tiny, but not very graceful.

```
--- a/include/kcl/backport/kcl_drm_backport.h
+++ b/include/kcl/backport/kcl_drm_backport.h
@@ -69,6 +69,12 @@ int _kcl_drm_syncobj_find_fence(struct drm_file *file_private,
 #define AMDKCL_ENABLE_RESIZE_FB_BAR
 #endif
 
+#ifdef OS_NAME_RHEL_7_X
+#ifndef AMDKCL_ENABLE_RESIZE_FB_BAR
+#define AMDKCL_ENABLE_RESIZE_FB_BAR
+#endif
+#endif
+
 /*
  * commit v4.10-rc3-539-g086f2e5cde74
  * drm: debugfs: Remove all files automatically on cleanup

```

2, We can get the OS_VERSION in m4 folder which to generate the configure, then define AMDKCL_ENABLE_RESIZE_FB_BAR when the OS_VERSION is centos7.x. But I think that is not a good idea.

Could you please provide me some better suggestions?

BR
Guo Lei

**Comment by fxkamd** [comment]
The way the tests in the m4 folder are written is, they try to compile a short snippet of code to try if it works or not. So you'd generate a short snippet of code that tries to call the kernel functions that resize the bar. If that succeeds, you can enable BAR resizing, otherwise you don't.

The idea is to eliminate any kernel version and OS version checks, which are very brittle, and replace them with actual feature availability checks that give the right answer regardless of the backports done to old kernels by OS vendors.

**Comment by raykwok1150** [comment]
Hi fxkamd ,

How about this patch?

```

gl@dcu:~/code/ROCm/ROCK-Kernel-Driver$ git diff --cached 
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu.h b/drivers/gpu/drm/amd/amdgpu/amdgpu.h
index 85ed160c7c81..a6e390403ee5 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu.h
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu.h
@@ -1219,7 +1219,7 @@ bool amdgpu_device_need_post(struct amdgpu_device *adev);
 
 void amdgpu_cs_report_moved_bytes(struct amdgpu_device *adev, u64 num_bytes,
                                  u64 num_vis_bytes);
-#ifdef AMDKCL_ENABLE_RESIZE_FB_BAR
+#ifdef HAVE_PCI_RESIZE_RESOURCE
 int amdgpu_device_resize_fb_bar(struct amdgpu_device *adev);
 #else
 static inline int amdgpu_device_resize_fb_bar(struct amdgpu_device *adev)
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_device.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_device.c
index 58ca17e894fa..64f7809bf086 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_device.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_device.c
@@ -925,7 +925,7 @@ void amdgpu_device_wb_free(struct amdgpu_device *adev, u32 wb)
                __clear_bit(wb, adev->wb.used);
 }
 
-#ifdef AMDKCL_ENABLE_RESIZE_FB_BAR
+#ifdef HAVE_PCI_RESIZE_RESOURCE
 /**
  * amdgpu_device_resize_fb_bar - try to resize FB BAR
  *
diff --git a/drivers/gpu/drm/amd/dkms/config/config.h b/drivers/gpu/drm/amd/dkms/config/config.h
index 94487e0dbc6b..7ac039511547 100644
--- a/drivers/gpu/drm/amd/dkms/config/config.h
+++ b/drivers/gpu/drm/amd/dkms/config/config.h
@@ -749,6 +749,9 @@
 /* pci_pcie_type() exist */
 #define HAVE_PCI_PCIE_TYPE 1
 
+/* pci_resize_resource() is available */
+#define HAVE_PCI_RESIZE_RESOURCE 1
+
 /* pci_upstream_bridge() is available */
 #define HAVE_PCI_UPSTREAM_BRIDGE 1
 
diff --git a/drivers/gpu/drm/amd/dkms/m4/kernel.m4 b/drivers/gpu/drm/amd/dkms/m4/kernel.m4
index c12b949c51a9..19fea941de2a 100644
--- a/drivers/gpu/drm/amd/dkms/m4/kernel.m4
+++ b/drivers/gpu/drm/amd/dkms/m4/kernel.m4
@@ -94,6 +94,7 @@ AC_DEFUN([AC_CONFIG_KERNEL], [
        AC_AMDGPU_PCI_UPSTREAM_BRIDGE
        AC_AMDGPU_PCIE_BANDWIDTH_AVAILABLE
        AC_AMDGPU_PCI_DEV_ID
+       AC_AMDGPU_PCI_RESIZE_RESOURCE
        AC_AMDGPU_PERF_EVENT_UPDATE_USERPAGE
        AC_AMDGPU_PTRACE_PARENT
        AC_AMDGPU_REGISTER_SHRINKER
diff --git a/drivers/gpu/drm/amd/dkms/m4/pci-resize-resource.m4 b/drivers/gpu/drm/amd/dkms/m4/pci-resize-resource.m4
new file mode 100644
index 000000000000..7a476fb0f168
--- /dev/null
+++ b/drivers/gpu/drm/amd/dkms/m4/pci-resize-resource.m4
@@ -0,0 +1,20 @@
+dnl #
+dnl # commit 8bb705e3e79d84e77edd4499e74483dd96a4626c
+dnl # Author: Christian König <christian.koenig@amd.com>
+dnl # Date:    Tue Oct 24 14:40:26 2017 -0500
+dnl # PCI: Add pci_resize_resource() for resizing BARs
+dnl #
+AC_DEFUN([AC_AMDGPU_PCI_RESIZE_RESOURCE], [
+       AC_KERNEL_DO_BACKGROUND([
+               AC_KERNEL_TRY_COMPILE([
+                       #include <linux/pci.h>
+               ], [
+                       int ret;
+                       pci_release_resource(NULL, 0);
+                       ret = pci_resize_resource(NULL, 0, 0);
+               ], [
+                       AC_DEFINE(HAVE_PCI_RESIZE_RESOURCE, 1,
+                               [pci_resize_resource() is available])
+               ])
+       ])
+])
diff --git a/include/kcl/backport/kcl_drm_backport.h b/include/kcl/backport/kcl_drm_backport.h
index f32e51d37d7d..191f8ff23ac3 100644
--- a/include/kcl/backport/kcl_drm_backport.h
+++ b/include/kcl/backport/kcl_drm_backport.h
@@ -61,14 +61,6 @@ int _kcl_drm_syncobj_find_fence(struct drm_file *file_private,
 #define drm_syncobj_find_fence _kcl_drm_syncobj_find_fence
 #endif
 
-/*
- * commit d3252ace0bc652a1a244455556b6a549f969bf99
- * PCI: Restore resized BAR state on resume
- */
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
-#define AMDKCL_ENABLE_RESIZE_FB_BAR
-#endif
-
 /*
  * commit v4.10-rc3-539-g086f2e5cde74
  * drm: debugfs: Remove all files automatically on cleanup
  */


```

BR
Guo Lei

**Comment by fxkamd** [comment]
Thank you, that looks very reasonable. Can you send that to me with "git send-email --to=felix.kuehling@amd.com"? Make sure you include a Signed-off-by: Your Name <your-addr@email.com> in the commit description. I'll work on getting it reviewed internally and included in an upcoming ROCm release.

**Comment by raykwok1150** [comment]
Hi, felix,
Done.

By the way, is there any script to genera dkms package from rock-kernel-driver?

BR
Guo Lei

**Comment by fxkamd** [comment]
I found out that a different fix for this issue has already been submitted internally. Unfortunately it didn't make it into the next release (ROCm 4.1). But this should make it into ROCm 4.2 later. I hope this will work with CentOS 7.9, i.e. the "PCI: Cleanup PCI_REBAR_CTRL_BAR_SHIFT handling" fix from 4.18 is in CentOS 7.9:

    drm/amdkcl: rework pci resize dependency.

    Reviewed-by: Guchun Chen <guchun.chen@amd.com>
    Signed-off-by: Flora Cui <flora.cui@amd.com>

diff --git a/include/kcl/backport/kcl_pci_backport.h b/include/kcl/backport/kcl_pci_backport.h
index 48098e386675..e6df88812236 100644
--- a/include/kcl/backport/kcl_pci_backport.h
+++ b/include/kcl/backport/kcl_pci_backport.h
@@ -15,10 +15,12 @@
 #endif

 /*
- * commit d3252ace0bc652a1a244455556b6a549f969bf99
- * PCI: Restore resized BAR state on resume
+ * v4.18-rc1-3-gb1277a226d8c PCI: Cleanup PCI_REBAR_CTRL_BAR_SHIFT handling
+ * v4.18-rc1-2-gd3252ace0bc6 PCI: Restore resized BAR state on resume
+ * v4.14-rc3-3-g8bb705e3e79d PCI: Add pci_resize_resource() for resizing BARs
+ * v4.14-rc3-2-g276b738deb5b PCI: Add resizable BAR infrastructure
  */
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+#ifdef PCI_REBAR_CTRL_BAR_SHIFT
 #define AMDKCL_ENABLE_RESIZE_FB_BAR
 #endif

**Comment by raykwok1150** [comment]
Well done!
Have a good weekend!
